File: e2e/typescript-orders/project/src/logging/logger.ts
/**
 * A small structured-logging abstraction. The service depends on the `Logger`
 * interface only, so callers can supply an in-memory recorder in tests or a
 * transport-backed implementation in production without touching domain code.
 */

export type LogLevel = "debug" | "info" | "warn" | "error";

export interface LogFields {
  readonly [key: string]: string | number | boolean;
}

export interface LogEntry {
  readonly level: LogLevel;
  readonly message: string;
  readonly fields: LogFields;
}

export interface Logger {
  log(entry: LogEntry): void;
}

/** Collects log entries in memory — the default for tests and I/O-free contexts. */
export class MemoryLogger implements Logger {
  readonly entries: LogEntry[] = [];

  log(entry: LogEntry): void {
    this.entries.push(entry);
  }
}

/** Emit an info-level entry through any logger. */
export function logInfo(logger: Logger, message: string, fields: LogFields = {}): void {
  logger.log({ level: "info", message, fields });
}

File: e2e/python-inventory/project/src/inventory/logging/logger.py
"""A small structured-logging abstraction.

The service depends on the :class:`Logger` protocol only, so callers can supply an
in-memory recorder in tests or a transport-backed logger in production without
touching domain code.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Protocol


@dataclass(frozen=True)
class LogEntry:
    """One structured log record."""

    level: str
    message: str
    fields: dict[str, object] = field(default_factory=dict)


class Logger(Protocol):
    def log(self, entry: LogEntry) -> None: ...


class MemoryLogger:
    """Collects log entries in memory — the default in tests and I/O-free code."""

    def __init__(self) -> None:
        self.entries: list[LogEntry] = []

    def log(self, entry: LogEntry) -> None:
        self.entries.append(entry)


def log_info(logger: Logger, message: str, **fields: object) -> None:
    """Emit an info-level entry through any logger."""
    logger.log(LogEntry(level="info", message=message, fields=dict(fields)))
